File: megakolmio.c
///////////////////////////////////////////////////////////////////////////////

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

///////////////////////////////////////////////////////////////////////////////

// NEIGHBORS:
// Puzzle is filled starting from position 0 to 8 in order. Placing first cards
// somewhere in the middle of the puzzle (as opposed to placing 0 on the top of
// the board) will greatly improve the performance since it yields less 
// solutions that end up non-completing late in the game.
// Below IDs are used as indexes to cardsOnBoard array.
//
//                 / \
//                /   \
//               /     \
//              /   6   \
//             /         \
//             -----------
//           / \         / \
//          /   \   0   /   \
//         /     \     /     \
//        /   2   \   /   1   \
//       /         \ /         \
//       ----------- -----------
//     / \         / \         / \
//    /   \   3   /   \   5   /   \
//   /     \     /     \     /     \
//  /   7   \   /   4   \   /   8   \
// /         \ /         \ /         \
// ----------- ----------- -----------

// EDGES:
// Below IDs are used to identify edges:
//
//   Up:             Down:
//       / \         -----------
//      /   \        \    2    /
//     /0   1\        \       /
//    /       \        \1   0/
//   /    2    \        \   /
//   -----------         \ /
//

// Neighbor relations and common edges:
#define NMAP_SIZE 9
#define NMAP_FIRST 0
#define NMAP_SECOND 1
#define NMAP_COMMON_EDGE 2
typedef unsigned char sint;
static const sint NEIGHBORMAP[NMAP_SIZE][3] = 
{
    { 0,  1,  0 },
    { 0,  2,  1 },
    { 0,  6,  2 },
    { 1,  5,  2 },
    { 2,  3,  2 },
    { 3,  4,  0 },
    { 3,  7,  1 },
    { 4,  5,  1 },
    { 5,  8,  0 } 
};

static const sint PRINTORDER[] = {6,2,0,1,7,3,4,5,8};

///////////////////////////////////////////////////////////////////////////////

#define CARD_NAME_LEN 2
#define EDGES_IN_CARD 3

typedef struct Card {
    const char name[CARD_NAME_LEN];
    const char* edges[EDGES_IN_CARD];
} Card;

///////////////////////////////////////////////////////////////////////////////

static const Card Deck[] = {
    "P1",{"FH","FB","DH"},
    "P2",{"DH","FB","RB"},
    "P3",{"DH","FB","FH"},
    "P4",{"DH","DB","FB"},
    "P5",{"DH","RB","DB"},
    "P6",{"RB","FB","RH"},
    "P7",{"FB","RH","FH"},
    "P8",{"RH","DH","RB"},
    "P9",{"FB","DB","DH"}
};

#define CARDS_IN_DECK (sizeof(Deck)/sizeof(Deck[0]))

///////////////////////////////////////////////////////////////////////////////

typedef struct PlayedCard {
    sint rotation;
    sint position;
    const Card* card;
} PlayedCard;

sint getCommonEdge(PlayedCard* this, PlayedCard* other) {
    for(sint i=0; i<NMAP_SIZE; i++) {
        if( NEIGHBORMAP[i][NMAP_FIRST]  == this->position &&
            NEIGHBORMAP[i][NMAP_SECOND] == other->position) {
            return NEIGHBORMAP[i][NMAP_COMMON_EDGE];
        }
    }
    printf("[+] ERROR: getCommonEdge(): not neighbors");
    exit(0);
}

bool matchesNeighbor(PlayedCard* this, PlayedCard* other) {
    if(this->card == NULL ) return false;
    sint common_edge = getCommonEdge(this, other);
    sint own_rotated_common_edge = (common_edge + this->rotation) % EDGES_IN_CARD;
    sint other_rotated_common_edge = (common_edge + other->rotation) % EDGES_IN_CARD;
    const char* e1 = this->card->edges[own_rotated_common_edge];
    const char* e2 = other->card->edges[other_rotated_common_edge];
    if(e1[0] == e2[0] && e1[1] != e2[1] ) {
        return true;
    }
    else {
        return false;
    }
}

bool rotate(PlayedCard* this) {
    if(this->rotation >= 2) {
        return false;
    }
    else {
        this->rotation += 1;
        return true;
    }
}

///////////////////////////////////////////////////////////////////////////////

typedef struct GameState {
    sint nextOnBoard;
    sint topOfTheDeck;
    PlayedCard* cardsOnBoard[CARDS_IN_DECK];
} GameState;

GameState* replicate(GameState* this) {
    GameState *newstate = calloc(1,sizeof(GameState));
    newstate->nextOnBoard = this->nextOnBoard;
    newstate->topOfTheDeck = this->topOfTheDeck;
    for(sint i=0; i < CARDS_IN_DECK; i++) {
        if(this->cardsOnBoard[i] == NULL) break;
        newstate->cardsOnBoard[i] = malloc(sizeof(PlayedCard));
        memcpy(newstate->cardsOnBoard[i],this->cardsOnBoard[i],sizeof(PlayedCard));
    }
    return newstate;
}

void delete(GameState* this) {
    for(sint i=0; i<CARDS_IN_DECK; i++) {
        if(this->cardsOnBoard[i]) {
            free(this->cardsOnBoard[i]);
        }
    }
    free(this);
}

bool isSolved(GameState* this, bool partial) {
    PlayedCard *c1, *c2;
    sint n1, n2;
    for(sint i=0; i<NMAP_SIZE; i++) {
        c1 = this->cardsOnBoard[NEIGHBORMAP[i][NMAP_FIRST]];
        c2 = this->cardsOnBoard[NEIGHBORMAP[i][NMAP_SECOND]];
        if (c1 == NULL || c2 == NULL) {
            if (partial) {
                continue;
            }
            else {
                return false;
            }
        }
        if (!matchesNeighbor(c1, c2)) {
            return false;
        }
    }
    return true;
}

void output(GameState* this) {
    printf("[");
    for(sint i=0; i<CARDS_IN_DECK; i++) {
        printf("%s", this->cardsOnBoard[PRINTORDER[i]]->card->name);
        if (i < CARDS_IN_DECK-1) printf(",");
    }
    printf("]\n");
}

bool isCardOnBoard(GameState* this, const Card* value) {
    for (sint i=0; i<CARDS_IN_DECK; ++i) {
        if (this->cardsOnBoard[i] != NULL && 
            this->cardsOnBoard[i]->card != NULL &&
            this->cardsOnBoard[i]->card == value) {
            return true;
        }
    }
    return false;
}

const Card* nextFromDeck(GameState* this) {
    if(this->topOfTheDeck >= CARDS_IN_DECK) {
        return NULL;
    }
    const Card* fromdeck = NULL;
    for(sint i=this->topOfTheDeck; i<CARDS_IN_DECK; i++) {
        fromdeck = &Deck[i];
        if(!isCardOnBoard(this, fromdeck)) {
            this->topOfTheDeck = i;
            return fromdeck;
        }

    }
    return NULL;
}

bool addNewCard(GameState* this) {
    const Card *fromdeck = nextFromDeck(this);
    if(fromdeck == NULL) {
        return false;
    }
    PlayedCard *newcard = malloc(sizeof(PlayedCard));
    newcard->position = this->nextOnBoard;
    newcard->rotation = 0;
    newcard->card = fromdeck;
    this->cardsOnBoard[newcard->position] = newcard;
    this->nextOnBoard++;
    return true;
}

bool replaceCard(GameState* this, PlayedCard *card) {
    const Card *fromdeck = nextFromDeck(this);
    if(fromdeck == NULL) {
        return false;
    }
    card->card = fromdeck;
    card->rotation = 0;
    return true;
}

PlayedCard* getLastAdded(GameState* this) {
    sint last = this->nextOnBoard - 1;
    return this->cardsOnBoard[last];
}

GameState* first(GameState* this) {
    GameState *newstate = replicate(this);
    newstate->topOfTheDeck = 0;
    if(!addNewCard(newstate)) {
        delete(newstate);
        return NULL;
    }
    return newstate;
}

GameState* next(GameState* this) {
    GameState *newstate = replicate(this);
    PlayedCard *card = getLastAdded(newstate);
    if (rotate(card)) {
        return newstate;
    }
    else if (!replaceCard(newstate, card)) {
        delete(newstate);
        return NULL;
    }
    return newstate;
}

///////////////////////////////////////////////////////////////////////////////

void solve(GameState* game) {
    if(!isSolved(game,true)) {
        return;
    }

    if(isSolved(game,false)) {
        output(game);
    }

    GameState *tempstate = NULL;
    GameState *newstate = first(game);
    while(newstate != NULL) {
        solve(newstate);
        tempstate = newstate;
        newstate = next(newstate);
        delete(tempstate);
    }
}

///////////////////////////////////////////////////////////////////////////////

int main() {
    GameState* game = calloc(1,sizeof(GameState));
    solve(game);
    delete(game);
}

///////////////////////////////////////////////////////////////////////////////

File: megakolmio.py
#!/usr/bin/env python

################################################################################

import sys
from copy import copy

################################################################################

# NEIGHBORS:
# Puzzle is filled starting from position 0 to 8 in order. Placing first cards
# somewhere in the middle of the puzzle (as opposed to placing 0 on the top of
# the board) will greatly improve the performance since it yields less solutions
# that end up non-completing late in the game.
# Below IDs are used as indexes to GameState.cardsOnBoard array.
#
#                 / \
#                /   \
#               /     \
#              /   6   \
#             /         \
#             -----------
#           / \         / \
#          /   \   0   /   \
#         /     \     /     \
#        /   2   \   /   1   \
#       /         \ /         \
#       ----------- -----------
#     / \         / \         / \
#    /   \   3   /   \   5   /   \
#   /     \     /     \     /     \
#  /   7   \   /   4   \   /   8   \
# /         \ /         \ /         \
# ----------- ----------- -----------

# EDGES:
# Below IDs are used to identify edges:
#
#   Up:             Down:
#       / \         -----------
#      /   \        \    2    /
#     /0   1\        \       /
#    /       \        \1   0/
#   /    2    \        \   /
#   -----------         \ /
#

# Neighbor relations and common edges:
NEIGHBORMAP = {}
NEIGHBORMAP[0,1] = 0
NEIGHBORMAP[0,2] = 1
NEIGHBORMAP[0,6] = 2
NEIGHBORMAP[1,5] = 2
NEIGHBORMAP[2,3] = 2
NEIGHBORMAP[3,4] = 0
NEIGHBORMAP[3,7] = 1
NEIGHBORMAP[4,5] = 1
NEIGHBORMAP[5,8] = 0    # cards on positions 5 and 8 have a common edge 0

# In the solution printout, the cards need to be printed from top to down and
# left to right:
PRINTORDER = [6,2,0,1,7,3,4,5,8]

################################################################################

class Deck:
    # Deck of cards: each card has a name and three edges
    def __init__(self):
        # F = Fox, D = Deer, R = Raccoon
        # H = Head, B = Body
        self.cards = []
        self.cards.append(Card("P1",["FH","FB","DH"]))
        self.cards.append(Card("P2",["DH","FB","RB"]))
        self.cards.append(Card("P3",["DH","FB","FH"]))
        self.cards.append(Card("P4",["DH","DB","FB"]))
        self.cards.append(Card("P5",["DH","RB","DB"]))
        self.cards.append(Card("P6",["RB","FB","RH"]))
        self.cards.append(Card("P7",["FB","RH","FH"]))
        self.cards.append(Card("P8",["RH","DH","RB"]))
        self.cards.append(Card("P9",["FB","DB","DH"]))
        self.nCardsInDeck = len(self.cards)

################################################################################

class Card:
    def __init__(self,name,edgenames):
        self.name = name
        self.edgeNames = edgenames

################################################################################

class PlayedCard:
    # PlayedCard is a specialized version of card - it includes two additional
    # properties: rotation and position. Position indicates the card's current
    # position on the board.
    # TODO: position in constructor
    # TODO: rotation with modulus
    def __init__(self,card):
        self.rotation = 0   # 0, 120, 240 degrees
        self.position = None
        self.card = card

    def matchesNeighbor(self,other):
        # Card matches another if body matches head, for instance:
        # FH (FoxHead) matches FB (FoxBody)
        if self.position == None:
            return False
        common_edge = NEIGHBORMAP[self.position,other.position]
        e1 = self.card.edgeNames[common_edge]
        e2 = other.card.edgeNames[common_edge]
        if e1[0] == e2[0] and e1[1] != e2[1]:
            return True
        else:
            return False

    def rotate(self):
        # Return False if the card would end up in it's original rotation
        if self.rotation >= 240:
            return False
        else:
            self.rotation += 120
            # Rotate edgeNames array: last becomes first
            self.card.edgeNames=[self.card.edgeNames[-1]]+self.card.edgeNames[0:-1]
            return True

################################################################################

class GameState():
    def __init__(self):
        # Next free position in cardsOnBoard array
        self.nextOnBoard = 0
        # Next card on the deck that will be next placed on the board
        self.topOfTheDeck = 0
        # Cards currently placed on the board
        self.cardsOnBoard = [None]*len(DECK.cards)
        # Card names currently on the board, we track this separately to
        # make it quicker to select the next card from the deck
        self.cardNamesOnBoard = [None]*len(DECK.cards)

    def isSolved(self, partial=False):
        # Return true if current solution is valid. If partial is True,
        # then consider also partial solutions valid.
        for c1_index,c2_index in NEIGHBORMAP:
            c1 = self.cardsOnBoard[c1_index]
            c2 = self.cardsOnBoard[c2_index]
            if None in [c1,c2]:
                if partial: continue
                return False
            if not c1.matchesNeighbor(c2):
                return False
        return True

    def output(self):
        # Output the solution in the specified PRINTORDER
        msg = []
        for i in PRINTORDER:
            msg.append(self.cardsOnBoard[i].card.name)
        print("[%s]" % ",".join(msg))

    def replicate(self):
        # Make a copy of the object
        new = GameState()
        new.nextOnBoard = self.nextOnBoard
        new.topOfTheDeck = self.topOfTheDeck
        new.cardsOnBoard = copy(self.cardsOnBoard)
        new.cardNamesOnBoard = copy(self.cardNamesOnBoard)
        return new

    def nextFromDeck(self):
        # Return the next card from the deck
        if self.topOfTheDeck >= DECK.nCardsInDeck:
            return None
        for i in range(self.topOfTheDeck, DECK.nCardsInDeck):
            fromdeck = DECK.cards[i]
            if fromdeck.name not in self.cardNamesOnBoard:
                self.topOfTheDeck = i
                return fromdeck
        return None

    def addNewCard(self):
        # Add a new card from the deck to the board
        fromdeck = self.nextFromDeck()
        if fromdeck == None:
            return False
        newcard = PlayedCard(fromdeck)
        newcard.position = self.nextOnBoard
        self.cardsOnBoard[newcard.position] = newcard
        self.cardNamesOnBoard[newcard.position] = newcard.card.name
        self.nextOnBoard += 1
        return True

    def replaceCard(self, card):
        # Replace the given card with a new card from the deck
        fromdeck = self.nextFromDeck()
        if fromdeck == None:
            return False
        card.card = fromdeck
        card.rotation = 0
        self.cardNamesOnBoard[card.position] = card.card.name
        return True

    def getLastAdded(self):
        # Return the card that was last added on the board
        last_added_index = self.nextOnBoard - 1
        return self.cardsOnBoard[last_added_index]

    def first(self):
        # Add next free card to the first empty place in the current
        # game state. Return the new game state if card was added on
        # the board. Otherwise return None.
        newstate = self.replicate()
        newstate.topOfTheDeck = 0
        if not newstate.addNewCard():
            return None
        return newstate

    def next(self):
        # If possible, try rotating the card added last on the board
        # in hope of finding a match. Otherwise select a new card.
        # Return new gamestate if card was added on the board. Otherwise
        # return None.
        newstate = self.replicate()
        card = newstate.getLastAdded()
        if card.rotate():
            return newstate
        elif not newstate.replaceCard(card):
            return None
        return newstate

################################################################################

# https://en.wikipedia.org/wiki/Backtracking
def solve(game):

    # The algorithm checks whether the game can be completed to a valid
    # solution. If it cannot, the whole sub-tree rooted at the given game state
    # is skipped (pruned).
    if not game.isSolved(partial=True):
        return

    # Otherwise, the algorithm first checks whether the current state is a valid
    # solution, and if so reports it to the user; ...
    if game.isSolved():
        game.output()


    # ... and recursively enumerates all sub-trees of current state
    newstate = game.first()
    while newstate != None:
        solve(newstate)
        newstate = newstate.next()

################################################################################

if __name__ == "__main__":
    DECK = Deck()
    game = GameState()
    solve(game)

################################################################################
